src/wp-includes/class-wp-block-scanner.php

```diff
@@ -0,0 +1,1380 @@
+	 * Scan to the next block delimiter in a document, indicating if one was found.
+	 *
+	 * Block comment delimiters must be valid HTML comments and may contain JSON.
+	 * This search does not determine, however, if the JSON is valid.
+	 *
+	 * Example delimiters:
+	 *
+	 *     `<!-- wp:paragraph {"dropCap": true} -->`
+	 *     `<!-- wp:separator /-->`
+	 *     `<!-- /wp:paragraph -->`
+	 *
+	 * In the case that a block comment delimiter contains both the void indicator and
+	 * also the closing indicator, it will be treated as a void block.
+	 *
+	 * Example:
+	 *
+	 *     // Find all image block opening delimiters.
+	 *     $images = array();
+	 *     $scanner = WP_Block_Scanner::create( $html );
+	 *     while ( $scanner->next_delimiter() ) {
+	 *         if ( $scanner->opens_block( 'core/image' ) ) {
+	 *             $images[] = $scanner->get_span();
+	 *         }
+	 *     }
+	 *
+	 * Not all blocks have explicit delimiters. Non-block content at the top-level of
+	 * a document (so-called “HTML soup”) forms implicit blocks containing neither a
+	 * block name nor block attributes. Because this content often comprises only
+	 * HTML whitespace and adds undo performance burden, it is skipped by default.
+	 * To scan the implicit freeform blocks, pass the `$html_spans` argument.
+	 *
+	 * Example:
+	 *
+	 *     $html    = '<!-- wp:void /-->\n<!-- wp:void /-->';
+	 *     $blocks  = [
+	 *       [ 'blockName' => 'core/void' ],
+	 *       [ 'blockName' => null ],
+	 *       [ 'blockName' => 'core/void' ],
+	 *     ];
+	 *     $scanner = WP_Block_Scanner::create( $html );
+	 *     while ( $scanner->next_delimiter( html_spans: 'visit' ) {
+	 *         ...
+	 *     }
+	 *
+	 * In some cases it may be useful to conditionally visit the implicit freeform
+	 * blocks, such as when determining if a post contains freeform content that
+	 * isn’t purely whitespace.
+	 *
+	 * Example:
+	 *
+	 *     $seen_block_types = [];
+	 *     $html_spans       = 'visit-html';
+	 *     $scanner          = WP_Block_Scanner::create( $html );
+	 *     while ( $scanner->next_delimiter( html_spans: $html_spans ) {
+	 *         if ( ! $scanner->opens_block() ) {
+	 *             continue;
+	 *         }
+	 *
+	 *         // Stop wasting time visiting freeform blocks after one has been found.
+	 *         if ('visit-html' === $html_spans ) {
+	 *             if ( $scanner->is_non_whitespace_freeform() ) {
+	 *                 $html_spans                        = 'skip-html';
+	 *                 $seen_block_types['core/freeform'] = true;
+	 *             }
+	 *             continue;
+	 *         }
+	 *
+	 *         $seen_block_types[ $scanner->get_block_type() ] = true;
+	 *     }
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param 'visit-html'|'skip-html' $html_spans Optional. Pass `visit-html` to match freeform HTML content
+	 *                                             not surrounded by block delimiters. Defaults to `skip-html`.
+	 * @return bool Whether a block delimiter was matched.
```

Comment:
What's the difference between a token and a delimiter? The PHPDoc is nearly identical here and for `next_delimiter()`.